modify eng system design (#936)
apache/incubator-iotdb: docs/Documentation/SystemDesign/5-DataQuery/5-GroupByQuery.md

<!--

    Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at
    
        http://www.apache.org/licenses/LICENSE-2.0
    
    Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

-->

# Downsampling query

* org.apache.iotdb.db.query.dataset.groupby.GroupByEngineDataSet

The result set of the downsampling query will inherit `GroupByEngineDataSet`, this class contains the following fields:
* protected long queryId
* private long interval
* private long slidingStep

> The following two fields are for the entire query, and the time period is left closed and right open, which is [startTime, endTime):

Reviewer: ```suggestion
The following two fields are for the entire query, and the time period is left closed and right open, which is `[startTime, endTime)`:
```

* private long startTime
* private long endTime

> The following fields are for the current segment.

Reviewer: ```suggestion
The following fields are for the current segment, and the time period is left closed and right open, which is `[startTime, endTime)`:
```

* protected long curStartTime;
* protected long curEndTime;
* private int usedIndex;
* protected boolean hasCachedTimeInterval;

> The core method of `GroupByEngineDataSet` is very easy. First, determine if there is a next segment based on whether there is a cached time period, and return` true`; if not, calculate the segmentation start time and increase `usedIndex` by 1.  If the segment start time has exceeded the query end time, return `false`; otherwise, calculate the query end time, set` hasCachedTimeInterval` to `true`, and return` true`:

Reviewer: ```suggestion
The core method of `GroupByEngineDataSet` is very easy. First, determine if there is a next segment based on whether there is a cached time period, and return `true`; if not, calculate the segmentation start time and increase `usedIndex` by 1.  If the segment start time has exceeded the query end time, return `false`; otherwise, calculate the query end time, set `hasCachedTimeInterval` to `true`, and return` true`:
```

```
protected boolean hasNextWithoutConstraint() {
  if (hasCachedTimeInterval) {
    return true;
  }

  curStartTime = usedIndex * slidingStep + startTime;
  usedIndex++;
  if (curStartTime < endTime) {
    hasCachedTimeInterval = true;
    curEndTime = Math.min(curStartTime + interval, endTime);
    return true;
  } else {
    return false;
  }
}
```

## Downsampling query without value filter

> The downsampling query logic without value filter is mainly in the `GroupByWithoutValueFilterDataSet` class, which inherits` GroupByEngineDataSet`.

Reviewer: ```suggestion
The downsampling query logic without value filter is mainly in the `GroupByWithoutValueFilterDataSet` class, which inherits `GroupByEngineDataSet`.
```

This class has the following key fields:
* private Map <Path, GroupByExecutor> pathExecutors classifies aggregate functions for the same `Path` and encapsulates them as` GroupByExecutor`,
  `GroupByExecutor` encapsulates the data calculation logic and method of each Path, which will be described later

* private TimeRange timeRange encapsulates the time interval of each calculation into an object, which is used to determine whether Statistics can directly participate in the calculation
* private Filter timeFilter Generates a user-defined query interval as a Filter object, which is used to filter the available files, chunks, and pages.

> First, in the initialization `initGroupBy ()` method, the `timeFilter` is calculated based on the expression, and` GroupByExecutor` is generated for each `path`.

Reviewer: ```suggestion
First, in the initialization `initGroupBy()` method, the `timeFilter` is calculated based on the expression, and `GroupByExecutor` is generated for each `path`.
```

> The `nextWithoutConstraint ()` method calculates the aggregate value `aggregateResults` of all aggregation methods in each` Path` by calling the `GroupByExecutor.calcResult ()` method.

Reviewer: ```suggestion
First, in the initialization `initGroupBy()` method, the `timeFilter` is calculated based on the expression, and `GroupByExecutor` is generated for each `path`.
```

        continue;
      }
    }
    // When Statistics is not available, you can only fetch all data for calculation
    BatchData batchData = reader.nextPage();
    if (batchData == null || !batchData.hasCurrent()) {
      continue;
    }
    // If the page just opened exceeds the time range, the data retrieved is cached and the calculation is directly ended.
    if (batchData.currentTime() >= curEndTime) {
      preCachedData = batchData;
      return true;
    }
    //Perform calculations
    calcFromBatch(batchData);
    ...
}

```

The `calcFromBatch` method is to traverse all the aggregate functions to calculate the retrieved BatchData. The main logic is:

```
for (Pair<AggregateResult, Integer> result : results) {
    //If a function has already been calculated, it will not be calculated, such as the minimum calculation.
    if (result.left.isCalculatedAggregationResult()) {
      continue;
    }
    // Perform calculations
    ....
}
//Determine if the data in the current batchdata can still be used next time, if it can be added to the cache
if (batchData.getMaxTimestamp() >= curEndTime) {
    preCachedData = batchData;
}
```

## Aggregated query with value filter

> The downsampling query logic with value filtering conditions is mainly in the GroupByWithValueFilterDataSet class, which inherits GroupByEngineDataSet.

Reviewer: ```suggestion
The downsampling query logic with value filtering conditions is mainly in the `GroupByWithValueFilterDataSet` class, which inherits `GroupByEngineDataSet`.
```

This class has the following key fields:
* private List<IReaderByTimestamp> allDataReaderList
* private GroupByPlan groupByPlan
* private TimeGenerator timestampGenerator
* private long timestamp is used to cache timestamp for the next group by partition
* private boolean hasCachedTimestamp used to determine whether there is a timestamp cache for the next group by partition
* private int timeStampFetchSize is the size of the group by calculating the batch

> First, in the initialization initGroupBy () method, create a timestampGenerator based on the expression; then create a SeriesReaderByTimestamp for each time series and place it in the allDataReaderList list. After initialization is complete, call the nextWithoutConstraint () method to update the result.  If timestamp is cached for the next group by partition and the time meets the requirements, add it to timestampArray, otherwise return the aggregateResultList result directly; if timestamp is not cached for the next group by partition, use timestampGenerator to traverse:

Reviewer: Add the `` around code-style words and pay attention to the space.

```
while (timestampGenerator.hasNext()) {
  // 调用 constructTimeArrayForOneCal() 方法，得到 timestamp 列表
  timeArrayLength = constructTimeArrayForOneCal(timestampArray, timeArrayLength);

  // 调用 updateResultUsingTimestamps() 方法，使用 timestamp 列表计算聚合结果
  for (int i = 0; i < paths.size(); i++) {
    aggregateResultList.get(i).updateResultUsingTimestamps(
        timestampArray, timeArrayLength, allDataReaderList.get(i));
  }

  timeArrayLength = 0;
  // 判断是否到结束
  if (timestamp >= curEndTime) {
    hasCachedTimestamp = true;
    break;
  }
}

> ```

Reviewer: Translate the Chinese in the comment
